src/Nethermind/Nethermind.Runner/Ethereum/Api/ApiBuilder.cs

```diff
@@ -24,18 +28,62 @@ public class ApiBuilder
     private readonly ILogManager _logManager;
     private readonly ILogger _logger;
     private readonly IInitConfig _initConfig;
+    private readonly IProcessExitSource _processExitSource;
     public ChainSpec ChainSpec { get; }
 
-    public ApiBuilder(IConfigProvider configProvider, ILogManager logManager)
+    public ApiBuilder(IProcessExitSource processExitSource, IConfigProvider configProvider, ILogManager logManager)
     {
         _logManager = logManager ?? throw new ArgumentNullException(nameof(logManager));
         _logger = _logManager.GetClassLogger();
+        _processExitSource = processExitSource;
         _configProvider = configProvider ?? throw new ArgumentNullException(nameof(configProvider));
         _initConfig = configProvider.GetConfig<IInitConfig>();
         _jsonSerializer = new EthereumJsonSerializer(configProvider.GetConfig<IJsonRpcConfig>().JsonSerializationMaxDepth);
         ChainSpec = LoadChainSpec(_jsonSerializer);
     }
 
+    public EthereumRunner CreateEthereumRunner(IEnumerable<INethermindPlugin> plugins)
+    {
+        bool wasCreated = Interlocked.CompareExchange(ref _apiCreated, 1, 0) == 1;
+        if (wasCreated)
+        {
+            throw new NotSupportedException("Creation of multiple APIs not supported.");
+        }
+
+        IEnumerable<IConsensusPlugin> consensusPlugins = plugins.OfType<IConsensusPlugin>();
+        if (consensusPlugins.Count() > 1)
+        {
+            throw new NotSupportedException($"More than one consensus plugins are enabled. {string.Join(", ", consensusPlugins.Select(x => x.Name))}");
+        }
+
+        ISpecProvider specProvider = new ChainSpecBasedSpecProvider(ChainSpec, _logManager);
+        FollowOtherMiners gasLimitCalculator = new FollowOtherMiners(specProvider);
+
+        IConsensusPlugin? enginePlugin = consensusPlugins.FirstOrDefault();
+        INethermindApi nethermindApi =
+            enginePlugin?.CreateApi(_configProvider, _jsonSerializer, _logManager, ChainSpec) ??
+            new NethermindApi(_configProvider, _jsonSerializer, _logManager, ChainSpec);
+        nethermindApi.SpecProvider = specProvider;
+        nethermindApi.GasLimitCalculator = gasLimitCalculator;
+        nethermindApi.ProcessExit = _processExitSource;
+        ((List<INethermindPlugin>)nethermindApi.Plugins).AddRange(plugins);
```

Comment:
This looks kind of off, can we design not to need this cast?

Reply:
Yeah I know, maybe there should exist some plugin collection that should be accessed separately, and only have read-only access through api?